storage/innobase/xtrabackup/src/xbcloud/http.cc

```diff
@@ -595,4 +597,79 @@ bool Http_client::make_async_request(const Http_request &request,
   return true;
 }
 
+template <typename CLIENT, typename CALLBACK>
+void Http_client::callback(CLIENT *client, std::string container,
+                           std::string name, Http_request *req,
+                           Http_response *resp, const Http_client *http_client,
+                           Event_handler *h, CALLBACK callback, CURLcode rc,
+                           const Http_connection *conn, ulong count) const {
+  bool retry_error = false;
+  if (rc == CURLE_OK && !resp->ok()) {
+    msg_ts("%s: Operation failed on object. Http error code: %lu\n",
+           my_progname, resp->http_code());
+    retry_error = true;
+  }
+
+  if (http_client->retriable_curl_error(rc)) {
+    retry_error = true;
+  } else if (rc != CURLE_OK && http_client->get_verbose()) {
+    msg_ts(
+        "%s: Curl error (%d) %s is not configured as retriable. You can allow "
+        "it by "
+        "adding --curl-retriable-errors=%d parameter\n",
+        my_progname, rc, curl_easy_strerror(rc), rc);
+  }
+  if (http_client->retriable_http_error(conn->response().http_code())) {
+    retry_error = true;
+  } else if (rc != CURLE_OK && http_client->get_verbose()) {
+    msg_ts(
+        "%s: http error (%lu) is not configured as retriable. You can allow it "
+        "by "
+        "adding --http-retriable-errors=%lu parameter\n",
+        my_progname, conn->response().http_code(),
+        conn->response().http_code());
+  }
+
+  if (retry_error && count <= client->get_max_retries()) {
+    ulong delay = get_exponential_backoff(count, client->get_max_backoff());
+    msg_ts("%s: Sleeping for %lu ns before retrying %s [%lu]\n", my_progname,
+           delay, name.c_str(), count);
+    std::this_thread::sleep_for(std::chrono::milliseconds(delay));
+    resp->reset_body();
+    client->signer->sign_request(client->hostname(container), container, *req,
+                                 time(0));
+    http_client->make_async_request(
+        *req, *resp, h,
+        std::bind(&Http_client::callback<CLIENT, CALLBACK>, this, client,
+                  container, name, req, resp, http_client, h, callback,
+                  std::placeholders::_1, std::placeholders::_2, count + 1),
```

Comment:
what do we pass for these place holders?

Reply:
This is later passed as https://github.com/percona/percona-xtrabackup/pull/1111/commits/9013bbaac5f6f63fb8d2fbad0dbc4b8c81dca454#diff-3e52c4cf4730dea8e8206bb9c0ca11545a0da2e035ed5bf4ca5389c74e970bd4R652

Reply:
Thanks!